keps/168-2-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,300 @@
+
+## Design Details
+
+The proposal introduces a new server running on the Kueue's pod. It computes the current state of KueueManager without any additional request overhead. The server uses the [K8s API Aggregation Layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) mechanism. The same mechanism is used by the [metrics-server](https://github.com/kubernetes-sigs/metrics-server). There will be no additional etcd objects or need to use existing ones. No additional requests to sync information across LocalQueues will be required. 
+
+Similarly to the ```metrics-server``` the server will be implemented with the [apiserver library](https://github.com/kubernetes/apiserver), which provides authentication and authorization.
+
+All computation will be done on-demand without additional reconcile loops.
+
+The server provides that:
+- Pending workloads are returned according to their actual status without significant delay. This includes adding and removing/admitting new workloads with various priorities,
+- Adding workloads to one LocalQueues results in position changes for workloads submitted to other LocalQueues,
+- Data is consistent across all LocalQueues,
+- User with only LocalQueue visibility cannot access the list of pending workloads for ClusterQueue.
+
+### API Details
+
+We introduce a new API that will extend the existing one.
+
+There will be separate endpoints for administrators and regular users. Each endpoint exposes information about a pending workload, such as:
+- workload's position in a ClusterQueue,
+- workload's position in a LocalQueue,
+- workload's priority,
+- creation timestamp.
+
+The API does not allow for the modification of any objects.
+
+Regular users will have access only to the LocalQueues they are assigned to. However, they will be able to fetch information about the global position of a workload in a ClusterQueue, without any details about workloads in different LocalQueues. 
+
+Administrators will have access to all the data at the ClusterQueue level. They will be able to view all the workloads, no matter the LocalQuees the workloads are assigned to.
+
+Since the K8s Aggregation Layer does not support pagination out of the box, we don't want to commit to it, since it would require significant effort. However, it can be implemented on the client side. We also expose [the endpoint to fetch information about a single workload](#fetch-information-about-a-single-workload), so there is a way to query its position without listing all the remaining pending workloads.
+
+### API endpoints:
+
+We introduce a new API group ```pending-workloads.kueue.x-k8s.io``` that aggregates the following endpoints:
+
+#### List all pending workloads in ClusterQueue
+
+```
+GET /apis/pending-workloads.kueue.x-k8s.io/VERSION/clusterqueues/CQ_NAME
```

Comment:
would something like this be possible?

```suggestion
GET /apis/visibility.kueue.x-k8s.io/VERSION/clusterqueues/CQ_NAME/pending_workloads
```

Reply:
We tried like this, but there seems to be a limitation of the framework, that the path can only have one fragment. The fragment is the key in this map in the prototype: https://github.com/kubernetes-sigs/kueue/blob/5bbb00f8c05ed02a84c11419e6af676222aaf787/pkg/visibility/api/install.go#L45-L47

Reply:
It seems that `/` is treated specially.
We do this for pods https://github.com/kubernetes/kubernetes/blob/029452198566a41bc39d04a1ec5bad3f37621a1c/pkg/registry/core/rest/storage_core.go#L231

Reply:
Thanks for pointing  that out @alculquicondor. Indeed, it's possible to introduce that kind of subresource so I'll change the KEP